File: frontend/src/api/requirements.ts
import { api } from "@/lib/api";

export type DocumentType =
  | "OIR"
  | "AIR"
  | "PIR"
  | "EIR"
  | "BEP"
  | "MIDP"
  | "TIDP"
  | "InformationProtocol"
  | "Supplementary";

export type DocumentStatus =
  "draft" | "under_review" | "approved" | "superseded" | "withdrawn";

export type ItemStatus = "not_met" | "partial" | "met";

export interface RequirementItem {
  id: string;
  document_id: string;
  item_no: string;
  what: string;
  when_required: string | null;
  how_required: string | null;
  for_whom: string | null;
  acceptance_criteria: string | null;
  responsible_user_id: string | null;
  status: ItemStatus;
  notes: string | null;
  due_date: string | null;
  milestone_name: string | null;
}

export interface RequirementsDocument {
  id: string;
  project_id: string;
  owner_user_id: string | null;
  doc_type: DocumentType;
  title: string;
  description: string | null;
  revision: string;
  status: DocumentStatus;
  effective_from: string | null;
  effective_to: string | null;
  created_at: string;
  updated_at: string;
  item_count: number;
  items: RequirementItem[];
}

export interface RequirementsDocumentCreate {
  doc_type: DocumentType;
  title: string;
  description?: string | null;
  revision?: string;
  status?: DocumentStatus;
}

export interface RequirementItemCreate {
  item_no: string;
  what: string;
  when_required?: string | null;
  how_required?: string | null;
  for_whom?: string | null;
  acceptance_criteria?: string | null;
  status?: ItemStatus;
  notes?: string | null;
  due_date?: string | null;
  milestone_name?: string | null;
}

export interface RequirementItemUpdate {
  item_no?: string;
  what?: string;
  when_required?: string | null;
  how_required?: string | null;
  for_whom?: string | null;
  acceptance_criteria?: string | null;
  status?: ItemStatus;
  notes?: string | null;
  due_date?: string | null;
  milestone_name?: string | null;
}

export interface RequirementsDocumentListResponse {
  items: RequirementsDocument[];
  total: number;
}

export const requirementsApi = {
  listDocuments: (
    projectId: string,
  ): Promise<RequirementsDocumentListResponse> =>
    api
      .get<RequirementsDocumentListResponse>(
        `/projects/${projectId}/requirements`,
      )
      .then((r) => r.data),

  getDocument: (
    projectId: string,
    docId: string,
  ): Promise<RequirementsDocument> =>
    api
      .get<RequirementsDocument>(`/projects/${projectId}/requirements/${docId}`)
      .then((r) => r.data),

  createDocument: (
    projectId: string,
    body: RequirementsDocumentCreate,
  ): Promise<RequirementsDocument> =>
    api
      .post<RequirementsDocument>(`/projects/${projectId}/requirements`, body)
      .then((r) => r.data),

  updateDocument: (
    projectId: string,
    docId: string,
    body: Partial<RequirementsDocumentCreate>,
  ): Promise<RequirementsDocument> =>
    api
      .patch<RequirementsDocument>(
        `/projects/${projectId}/requirements/${docId}`,
        body,
      )
      .then((r) => r.data),

  deleteDocument: (projectId: string, docId: string): Promise<void> =>
    api
      .delete(`/projects/${projectId}/requirements/${docId}`)
      .then(() => undefined),

  createItem: (
    projectId: string,
    docId: string,
    body: RequirementItemCreate,
  ): Promise<RequirementItem> =>
    api
      .post<RequirementItem>(
        `/projects/${projectId}/requirements/${docId}/items`,
        body,
      )
      .then((r) => r.data),

  updateItem: (
    projectId: string,
    docId: string,
    itemId: string,
    body: RequirementItemUpdate,
  ): Promise<RequirementItem> =>
    api
      .patch<RequirementItem>(
        `/projects/${projectId}/requirements/${docId}/items/${itemId}`,
        body,
      )
      .then((r) => r.data),

  deleteItem: (
    projectId: string,
    docId: string,
    itemId: string,
  ): Promise<void> =>
    api
      .delete(`/projects/${projectId}/requirements/${docId}/items/${itemId}`)
      .then(() => undefined),
};

File: backend/app/schemas/requirements.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

DocumentTypeValues = Literal[
    "OIR",
    "AIR",
    "PIR",
    "EIR",
    "BEP",
    "MIDP",
    "TIDP",
    "InformationProtocol",
    "Supplementary",
]
DocumentStatusValues = Literal[
    "draft", "under_review", "approved", "superseded", "withdrawn"
]
ItemStatusValues = Literal["met", "partial", "not_met"]


class RequirementItemCreate(BaseModel):
    item_no: str
    what: str
    when_required: str | None = None
    how_required: str | None = None
    for_whom: str | None = None
    acceptance_criteria: str | None = None
    responsible_user_id: str | None = None
    status: ItemStatusValues = "not_met"
    notes: str | None = None
    due_date: str | None = None
    milestone_name: str | None = None


class RequirementItemUpdate(BaseModel):
    item_no: str | None = None
    what: str | None = None
    when_required: str | None = None
    how_required: str | None = None
    for_whom: str | None = None
    acceptance_criteria: str | None = None
    responsible_user_id: str | None = None
    status: ItemStatusValues | None = None
    notes: str | None = None
    due_date: str | None = None
    milestone_name: str | None = None


class RequirementItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    document_id: str
    item_no: str
    what: str
    when_required: str | None
    how_required: str | None
    for_whom: str | None
    acceptance_criteria: str | None
    responsible_user_id: str | None
    status: str
    notes: str | None
    due_date: str | None
    milestone_name: str | None


class RequirementsDocumentCreate(BaseModel):
    doc_type: DocumentTypeValues
    title: str
    revision: str = "01"
    status: DocumentStatusValues = "draft"
    effective_from: str | None = None
    effective_to: str | None = None
    description: str | None = None


class RequirementsDocumentUpdate(BaseModel):
    title: str | None = None
    revision: str | None = None
    status: DocumentStatusValues | None = None
    effective_from: str | None = None
    effective_to: str | None = None
    description: str | None = None


class RequirementsDocumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    project_id: str
    owner_user_id: str | None
    doc_type: str
    title: str
    revision: str
    status: str
    effective_from: str | None
    effective_to: str | None
    description: str | None
    created_at: datetime
    updated_at: datetime
    item_count: int
    items: list[RequirementItemResponse]


class RequirementsDocumentListResponse(BaseModel):
    items: list[RequirementsDocumentResponse]
    total: int
